crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2359,12 +2361,24 @@ impl<'db> TypeInferenceBuilder<'db> {
 
             // No nonlocal binding, check module globals. Avoid infinite recursion if `self.scope`
             // already is module globals.
-            let ty = if file_scope_id.is_global() {
-                Type::Unbound
-            } else {
-                global_symbol_ty(self.db, self.file, name)
+            if !file_scope_id.is_global() {
+                ty = global_symbol_ty(self.db, self.file, name);
             };
 
+            // Still possibly unbound? All modules are instances of `types.ModuleType`;
+            // look it up there (with a few very special exceptions)
+            if ty.may_be_unbound(self.db)
+                && !matches!(&**name, "__dict__" | "__init__" | "__getattr__")
```

Comment:
Hmmm... since we anyway having to hardcode a denylist of attributes _not_ to lookup on `types.ModuleType`, I suppose I could instead have an allowlist of attributes we _should_ lookup on `types.ModuleType`.

That has the downside that new implicit globals might be added to the `types.ModuleType` stub in the future and we wouldn't automatically recognise them as such unless we updated our hardcoded allowlist. But it might be much better for performance.

Reply:
Interestingly, the regression is for the incremental benchmark. The cold benchmark is barely affected. Not sure why that is. Some local profiling could help to understand the difference. 

To me the main question here is: How common is that we reach unbound here? If it's not very common in valid code (unless we expect specific attributes), than the performance penalty in actual code should be low.

Reply:
One issue is that we have to lookup symbols on `ModuleType` before we look them up in the builtins scope. But it's realistically much more likely that a symbol is going to resolve to a symbol in `builtins` than that it's going to be an attribute on `ModuleType` :/ So I think it's quite common that we'd reach `Unbound` here and then have to continue to the next part of the algorithm where we look it up in `builtins.pyi`.

Reply:
True. But let's only do that if this is the root cause

Reply:
Sure -- but I would strongly suspect that the root cause has _something_ to do with looking up `ModuleType` in typeshed and looking up a member on it, since that's really the only change in this diff. Currently this diff does that work for every lookup of a builtin (which are common), we could instead do it only for those names where it is relevant.

So my next steps here would be to a) implement a hardcoded version of the name allowlist (because it's quick and easy) and see if that fixes the regression (I strongly suspect it will), and if so, then turn it into a Salsa query allowlist instead, so we remain robust to typeshed changes.

Reply:
Confirmed that hardcoding an allowlist rather than a denylist does fix the perf regression!